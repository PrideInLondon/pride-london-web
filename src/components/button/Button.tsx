import React from 'react'
import { handleUrl } from '../../utilities'
import { StyledButton } from './Button.styles'
import {
  ButtonAsAnchorProps,
  ButtonProps,
  ButtonSignature,
} from './Button.types'

/**
 * Guard to check if `to` exists in props
 */
export const hasToProp = (
  props: ButtonProps | Omit<ButtonAsAnchorProps, 'href'>
): props is Omit<ButtonAsAnchorProps, 'href'> => 'to' in props

export const Button: ButtonSignature = (
  props: Omit<ButtonAsAnchorProps, 'href'> | ButtonProps
) => {
  const { to, children } = props

  if (hasToProp(props)) {
    const { as, ...anchorProps } = handleUrl(to!)
    return (
      <StyledButton renderAs={as} {...anchorProps} {...props}>
        {children}
      </StyledButton>
    )
  }
  return (
    <StyledButton renderAs="button" {...props}>
      {children}
    </StyledButton>
  )
}

Button.defaultProps = {
  variant: 'primary',
}
