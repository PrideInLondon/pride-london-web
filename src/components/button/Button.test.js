import React from 'react'
import { shallow, mount } from 'enzyme'
import Button from './'

describe('<Button/>', () => {
  it('matches snapshot', () => {
    const wrapper = shallow(<Button />)
    expect(wrapper).toMatchSnapshot()
  })

  it('fires correct method when set onClick property', () => {
    let counter = 0
    const handleClick = () => (counter += 1)
    const wrapper = shallow(<Button onClick={handleClick} />)
    wrapper.prop('onClick')()
    expect(counter).toBe(1)
  })

  it('renders a gatsby link if the to prop is not http', () => {
    const wrapper = mount(<Button to="/events" />)
    expect(wrapper.find('GatsbyLink').exists()).toBeTruthy()
  })

  it('renders a normal a tag if the to prop is https', () => {
    const wrapper = mount(<Button to="https://google.com" />)
    expect(wrapper.find('a')).toBeTruthy()
    expect(wrapper.find('GatsbyLink').exists()).toBeFalsy()
  })

  it('renders a button tag if there is no to prop', () => {
    const wrapper = mount(<Button />)
    expect(wrapper.find('button')).toBeTruthy()
  })
})
