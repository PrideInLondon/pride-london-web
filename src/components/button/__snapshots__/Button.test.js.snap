// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button/> matches snapshot 1`] = `
<Buttonstyles__StyledButton
  renderAs="button"
  variant="primary"
/>
`;
